### PR TITLE
Restructure locations prepared to use snapshot date

### PR DESCRIPTION
**Pull request by AdamProbert**
Ticket: https://trello.com/c/H3X2hJZr/174-epic2-restructure-locations-prepared-to-handle-returning-all-cqc-ascwds-data

Problem: 
- The prepare_locations dataset only has cqc locations for 2021
- We need locations prepared for all available years (2012+)

This ticket does the following: 
- Determines all unique import_dates for cqc locations, cqc providers, cqc pir and asc workplaces.
- Builds out a `date_matrix` identifying for each asc import which imports of the other datasets are most relevant.
- Restructures the main function to iterate over this date matrix building out a dataset utilising these `date_matrix` rows.
- Adds much more testing utilising test data generators instead of static files.
- Adds utility for accessing environmental variables

Example athena query for viewing new dataset:
```
select 
    * 
from 
    dataset_locations_prepared 
where 
    version = '1.1.5'
order by
    locationid;
 ```

### CODE REVIEW

File: tests/unit/test_prepare_locations.py [hunk 1]
```diff
@@ -1,7 +1,16 @@
...
 import unittest
+import mock
+import os
 from pathlib import Path
+from tests.test_file_generator import (
+    generate_ascwds_workplace_file,
+    generate_cqc_locations_file,
+    generate_cqc_providers_file,
+    generate_pir_file,
+)
```
[the remark this answers is not in the document]
Comment (AdamProbert) [reply tests/unit/test_prepare_locations.py:14]: Aye good catch thanks 👍

File: tests/unit/test_prepare_locations.py [hunk 2]
```diff
@@ -13,10 +22,18 @@
 )
 
 from jobs import format_fields, prepare_locations
```
[the remark this answers is not in the document]
Comment (AdamProbert) [reply tests/unit/test_prepare_locations.py:24]: 🖖

File: tests/unit/test_prepare_locations.py [hunk 3]
```diff
@@ -27,54 +44,177 @@ class PrepareLocationsTests(unittest.TestCase):
...
+            pass  # Ignore dir does not exist
+
+    def test_main_successfully_runs(self):
+        output_df = prepare_locations.main(
+            self.TEST_ASCWDS_WORKPLACE_FILE,
+            self.TEST_CQC_LOCATION_FILE,
+            self.TEST_CQC_PROVIDERS_FILE,
+            self.TEST_PIR_FILE,
+        )
+        output_df.select("snapshot_date", "cqc_pir_import_date").show()
```
[the remark this answers is not in the document]
Comment (AdamProbert) [reply tests/unit/test_prepare_locations.py:74]: Certainly don't 💯

File: tests/unit/test_prepare_locations.py [hunk 4]
```diff
@@ -27,54 +44,177 @@ class PrepareLocationsTests(unittest.TestCase):
...
+        result = prepare_locations.get_date_closest_to_search_date(test_date, date_search_list)
+        self.assertEqual(result, None)
+
+    def test_get_date_closest_to_search_date_with_single_valid_date_returns_date(self):
+        date_search_list = [date(2020, 3, 31)]
+        test_date = date(2022, 5, 1)
+
+        result = prepare_locations.get_date_closest_to_search_date(test_date, date_search_list)
+        self.assertEqual(result, date(2020, 3, 31))
+
```
[the remark this answers is not in the document]
Comment (JimmySunDance) [reply tests/unit/test_prepare_locations.py:160]: Not sure why this comment has doubled up

**Review by JimmySunDance — APPROVED** [review]
Looks good to me! Good work!

**Review by garypolzin — APPROVED** [review]
looks great

I'm just wondering if, instead of adding ASCWDS data as a left join, we add in as a full outer join matched on locationid?

That way we always have access to all CQC data and all ASCWDS data, plus the locationid link between the two

It won't break 'estimate_2021_jobs' as the initial filter` f"{REGISTRATION_STATUS} = 'Registered'` will wipe the ASCWDS accounts who aren't matched on locationid out

**Pull request merged**